File: functions/Leon.m
classdef Leon < FunctionND
  %
  % The Leon cubic valley function.
  %
  %   based on a code of John Burkardt (http://people.sc.fsu.edu/~jburkardt/)
  %
  % Reference:
  %
  %   Richard Brent,
  %   Algorithms for Minimization with Derivatives,
  %   Dover, 2002,
  %   ISBN: 0-486-41998-3,
  %   LC: QA402.5.B74.
  %
  % Author:
  %   Enrico Bertolazzi
  %   Dipartimento di Ingegneria Indutriale
  %   Universita` degli Studi di Trento
  %   email: enrico.bertolazzi@unitn.it
  %

  methods

    function self = Leon()
      self@FunctionND(int32(2)) ;
      self.exact_solutions = [ 1.0; 1.0  ];     % one known solution 
      self.guesses         = [ -1.2; -1.0 ];
    end

    function f = eval(self,x)
      % evaluate function
      self.check_x(x);
      f1 = x(2) - x(1) * x(1) * x(1);
      f2 = 1.0 - x(1);
      f  = 100.0 * f1 * f1 + f2 * f2;
    end

    function g = grad( self, x )
      % use analitic gradient
      self.check_x(x);
      g    = zeros ( 1, 2 );
      g(1) = - 600.0 * ( x(2) - x(1)^3 ) * x(1) * x(1) - 2.0 * ( 1.0 - x(1) );
      g(2) = 200.0 * ( x(2) - x(1)^3 );
    end

    function h = hessian( self, x )
      % use analitic hessian
      self.check_x(x);
      h = zeros ( 2, 2 );
      h(1,1) = - 1200.0 * x(1) * x(2) + 3000.0 * x(1)^4 + 2.0;
      h(1,2) = - 600.0 * x(1) * x(1);
      h(2,1) = - 600.0 * x(1) * x(1);
      h(2,2) = 200.0;
    end
  end
end

File: julia/functions/Leon.jl
include("../libj/FunctionND.jl")
N = 2
M = 1

function Leon()
    return x -> 100.0*(x[2] - x[1] * x[1] * x[1])^2 + (1.0 - x[1])^2
end

function evalf( x )
      # evaluate function
      check_x(x)
      f1 = x[2] - x[1] * x[1] * x[1]
      f2 = 1.0 - x[1]
      return (100.0 * f1 * f1 + f2 * f2)
end

function grad( x )
      # use analitic gradient
      check_x(x)
      g    = zeros( Float64, 1, 2 )
      g[1] = - 600.0 * ( x[2] - x[1]^3 ) * x[1] * x[1] - 2.0 * ( 1.0 - x[1] )
      g[2] = 200.0 * ( x[2] - x[1]^3 )
      return g
end

function hessian( x )
      # use analitic hessian
      check_x(x)
      h = zeros( Float64, 2, 2 )
      h[1,1] = - 1200.0 * x[1] * x[2] + 3000.0 * x[1]^4 + 2.0
      h[1,2] = - 600.0 * x[1] * x[1]
      h[2,1] = - 600.0 * x[1] * x[1]
      h[2,2] = 200.0
      return h
end
